src/_pytest/subtests.py

```diff
@@ -0,0 +1,431 @@
+class _SubTestContextManager:
+    """
+    Context manager for subtests, capturing exceptions raised inside the subtest scope and handling
+    them through the pytest machinery.
+    """
+
+    # Note: initially the logic for this context manager was implemented directly
+    # in Subtests.test() as a @contextmanager, however, it is not possible to control the output fully when
+    # exiting from it due to an exception when in `--exitfirst` mode, so this was refactored into an
+    # explicit context manager class (pytest-dev/pytest-subtests#134).
+
+    ihook: pluggy.HookRelay
+    msg: str | None
+    kwargs: dict[str, Any]
+    suspend_capture_ctx: Callable[[], AbstractContextManager[None]]
+    request: SubRequest
+    reporter: SubtestsReporterPlugin
+
+    def __enter__(self) -> None:
+        __tracebackhide__ = True
+
+        self._start = time.time()
+        self._precise_start = time.perf_counter()
+        self._exc_info = None
+
+        self._exit_stack = ExitStack()
+        self._captured_output = self._exit_stack.enter_context(
+            capturing_output(self.request)
+        )
+        self._captured_logs = self._exit_stack.enter_context(
+            capturing_logs(self.request)
+        )
+
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_val: BaseException | None,
+        exc_tb: TracebackType | None,
+    ) -> bool:
+        __tracebackhide__ = True
+        try:
+            if exc_val is not None:
+                exc_info = ExceptionInfo.from_exception(exc_val)
+            else:
+                exc_info = None
+        finally:
+            self._exit_stack.close()
+
+        precise_stop = time.perf_counter()
+        duration = precise_stop - self._precise_start
+        stop = time.time()
+
+        call_info = CallInfo[None](
+            None,
+            exc_info,
+            start=self._start,
+            stop=stop,
+            duration=duration,
+            when="call",
+            _ispytest=True,
+        )
+        report = self.ihook.pytest_runtest_makereport(
+            item=self.request.node, call=call_info
+        )
+        sub_report = SubtestReport._from_test_report(
+            report, SubtestContext(msg=self.msg, kwargs=self.kwargs.copy())
+        )
+
+        if sub_report.failed:
+            self.reporter.contains_failed_subtests[self.request.node.nodeid] += 1
+
+        self._captured_output.update_report(sub_report)
+        self._captured_logs.update_report(sub_report)
+
+        with self.suspend_capture_ctx():
+            self.ihook.pytest_runtest_logreport(report=sub_report)
+
+        if check_interactive_exception(call_info, sub_report):
+            self.ihook.pytest_exception_interact(
+                node=self.request.node, call=call_info, report=sub_report
+            )
+
+        if exc_val is not None:
+            if self.request.session.shouldfail:
+                return False
+        return True
```

Comment:
There's a complication here, this misses the `reraise` handling that `CallInfo.from_call` has:

```py
    reraise: tuple[type[BaseException], ...] = (Exit,)
    if not item.config.getoption("usepdb", False):
        reraise += (KeyboardInterrupt,)
```

So `subtests.test()` ends up swallowing `pytest.exit()` and Ctrl-C. Can test it with:

```py
import pytest

def test(subtests):
    with subtests.test():
        pytest.exit() # To test exit
        # import time; time.sleep(5) # To test Ctrl-C
    print('AFTER')

def test2(): pass
```

There's also a question of `pytest.skip()` behavior. Currently it skips the subtest. Maybe that's what we want? Didn't think about it too much.

Reply:
Good catch, fixed!